bundle/bundle-writer.go

```diff
@@ -0,0 +1,352 @@
+	if err != nil {
+		return err
+	}
+
+	manifestFile, err = addDatabase(manifestFile, databases)
+
+	manifestFile, err = getRequiredStencils(
+		templateRepository,
+		branch,
+		outputDir,
+		services,
+		skycapFolder,
+		manifestFile,
+		githubURL)
+
+	if err != nil {
+		return err
+	}
+
+	manifestFile, err = addMetadatas(manifestFile)
+
+	if err != nil {
+		return err
+	}
+
+	err = saveManifest(skycapFolder, manifestFile)
+	if err !=nil {
+		return err
+	}
+	return err
+}
+
+// downloading templates from github and putting them into homedir
+func getStencilTemplateFile(templateRepository string, skycapFolder string, filename string, branch string) (string, error) {
+
+	//Download templates.json file
+	manifest_path := templateRepository + filename // don't need to use filepath since it's a URL
+	temp_folder := filepath.Join(skycapFolder, "temp")
+	down_err := common.DownloadSingleFile(temp_folder, common.DownloadFile{URL: manifest_path, Name: filename}, branch)
+	if down_err != nil {
+		return "", down_err
+	}
+	return filepath.Join(temp_folder, filename), nil
+}
+
+func getEnvVars (servs []*common.Service, databases []common.Database) (map[string]string) {
+	var envas = make(map[string]string)
+	for _, envVarArray := range servs {
+		for _, envs := range envVarArray.EnvVars {
+			envas[envs.Key] = envs.Value
+		}
+	}
+	return envas
+}
+
+func createBundleFolderStructure(baseFolder string) error {
+	var folders = [6]string{"stencils", "policies", "stencil-group", "helm-releases", "temp", "configurations"}
+	for _, subfolder := range folders {
+		folder := filepath.Join(baseFolder, subfolder)
+		err := os.MkdirAll(folder, 0777)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func getRequiredStencils(templateRepository string,
+						branch string,
+						outputDir string,
+						services []*common.Service,
+						skycapFolder string,
+						manifestFile ManifestBundle,
+						githubURL string) 	(ManifestBundle, error){
+
+	//start download the template.json file
+	tjPathfile, err := getStencilTemplateFile(templateRepository, skycapFolder, "templates.json", branch)
+	if err != nil {
+		fmt.Printf("Error while downloading the templates.json. err: %s", err)
+		return ManifestBundle{}, err
+	}
+	// open the template.json file and start downloading the stencils
+	templateJson, err := os.Open(tjPathfile)
+	if err != nil {
+		return ManifestBundle{}, err
+	}
+
+	templatesJsonData, err := ioutil.ReadAll(templateJson)
+	if err != nil {
+		return ManifestBundle{}, err
+	}
+
+	var templJson map[string]interface{}
+	err = json.Unmarshal([]byte(templatesJsonData), &templJson)
+	if err != nil {
+		return ManifestBundle{}, err
+	}
+
+	var manifestStencils = make([]*BundleStencil, 0)
+
+	for i, data := range templJson {
+		if i =="templates" {
+			for _, stencils := range data.([]interface{}) {
+				stencil := stencils.(map[string]interface{})
+				if stencil["min_usage"].(float64) > 0 {
```

Comment:
this might be ok here if we have a validation pre-step somewhere else.

Reply:
I will create the needed structs.